miniscope_io/bit_operation.py

```diff
@@ -0,0 +1,93 @@
+"""
+Bit operations for parsing header and payload information. Currently for use in streamDaq module
+"""
+
+from typing import Tuple
+
+import numpy as np
+
+
+class BufferFormatter:
+    """
+    A class for formatting buffers and returning header (uint32) and payload (uint8) data.
+    """
+
+    @staticmethod
+    def _reverse_bits_in_array(arr: np.ndarray) -> np.ndarray:
+        """
+        Reverse the bits in each element of a numpy array.
+
+        Args:
+            arr (np.ndarray): The input array.
+
+        Returns:
+            np.ndarray: The array with bits in each element reversed.
+        """
+        arr = np.unpackbits(arr.view(np.uint8)).reshape(-1, 32)[:, ::-1]
+        return np.packbits(arr).view(np.uint32)
+
+    @staticmethod
+    def _reverse_byte_order_in_array(arr: np.ndarray) -> np.ndarray:
+        """
+        Reverse the byte order of each element in a numpy array.
+
+        Args:
+            arr (np.ndarray): The input array.
+
+        Returns:
+            np.ndarray: The array with byte order in each element reversed.
+        """
+        return arr.byteswap()
+
+    @classmethod
+    def bytebuffer_to_ndarrays(
+        cls,
+        buffer: bytes,
+        header_length_words: int,
+        preamble_length_words: int,
+        reverse_header_bits: bool = True,
+        reverse_header_bytes: bool = True,
+        reverse_body_bits: bool = True,
+        reverse_body_bytes: bool = True,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Format the buffer by optionally reversing bits in the header and body.
+
+        Args:
+            buffer (bytes): The binary buffer to format.
+            header_length_words (int): Length of the header in words.
+            preamble_length_words (int): Length of the preamble in words.
+            reverse_header_bits (bool): Reverse bits in the header if True. Default is True.
+            reverse_header_bytes (bool): Reverse byte order in the header if True. Default is True.
+            reverse_body_bits (bool): Reverse bits in the body if True. Default is True.
+            reverse_body_bytes (bool): Reverse byte order in the body if True. Default is True.
+
+        Returns:
+            tuple: A tuple containing the processed header as a numpy array of uint32
+                   and the processed payload as a numpy array of uint8.
+        """
+
+        # Padding in case the buffer is not a multiple of 4 bytes (32 bits)
+        padding_length = (4 - (len(buffer) % 4)) % 4
+        padded_buffer = buffer + b"\x00" * padding_length
+
+        # Convert the padded buffer to a uint32 numpy array
+        data = np.frombuffer(padded_buffer, dtype=np.uint32)
```

Comment:
This comes from how we format headers (metadata) in the microcontroller. Each pixel we want to transfer here is 8-bit, but that's too small for metadata, so we use 32-bit.

I'll add these to the docs, but the buffer is a concat of a 32-bit array for metadata and an 8-bit array for payload. We want these bit operations because we treat this as a 32-bit array in the microcontroller, making it confusing to do all bit/byte ordering before sending.